docs/proposals/parquet-storage.md

```diff
@@ -0,0 +1,139 @@
+   - Initial sync causes long startup time. This affects service availability on both scaling and failure recovery scenario
+
+3. Query Inefficiencies
+   - Attempts to minimize storage requests often lead to overfetching, causing high bandwidth usage
+   - Complex caching logic with varying effectiveness. Latency varies a lot when cache miss
+   - Processes single block with one goroutine, leading to high latency for large blocks and cannot scale without complex data partitioning
+
+### Why Parquet?
+
+[Apache Parquet](https://parquet.apache.org/) is a columnar storage format designed specifically for efficient data storage and retrieval from object storage systems. It offers several key advantages that directly address the problems we face with TSDB and Store Gateway:
+
+- Data organized by columns rather than rows, reduces number of requests to object storage as only limited IO is required to fetch the whole column
+- Rich file metadata and index, no local state like index header required to query the data, making it stateless
+- Advanced compression techniques reduce storage costs and improve query performance
+- Parallel processing friendly using Parquet Row Group
+
+There are other benefits of Parquet formats, but they are not directly related to the proposal:
+
+- Wide ecosystem and tooling support
+- Column pruning opportunity using projection pushdown
+
+## Out of Scope
+
+- Allow Ingester and Compactor to create Parquet files instead of TSDB blocks directly. This could be in the future roadmap but this proposal only focuses on converting and querying Parquet files.
+
+## Proposed Design
+
+### Components
+
+There are 2 new Cortex components/modules introduced in this design.
+
+#### 1. Parquet Converter
+
+Parquet Converter is a new component that converts TSDB blocks on object store to Parquet file format.
+
+It is similar to compactor, however, it only converts single block. The converted Parquet files will be stored in the same TSDB block folder so that the lifecycle of Parquet file will be managed together with the block.
+
+Only certain blocks can be configured to convert to Parquet file and it can be block duration based, for example we only convert if block duration is >= 12h.
+
+#### 2. Parquet Queryable
+
+Similar to the existing `distributorQueryable` and `blockStorageQueryable`, Parquet queryable is a queryable implementation which allows Cortex to query parquet files and can be used in both Cortex Querier and Ruler.
+
+If Parquet queryable is enabled, block storage queryable will be disabled and Cortex querier will not query Store Gateway anymore. `distributorQueryable` remains unchanged so it still queries Ingesters.
+
+Parquet queryable uses bucket index to discovers  parquet files in object storage. The bucket index is the same as the existing TSDB bucket index file, but using a different name `bucket-index-parquet.json.gz`. It is updated periodically by Cortex Compactor/Parquet Converter if parquet storage is enabled.
```

Comment:
How do we still query the 12h tsdb blocks while we dont have them in the parquet index?
For example if we have 8h blocks and we are compacting to 12h blocks, I would assume after the new 12h tsdb blocks are created we convert it to parquet, but while this doesnt finish we only have them in the default index?

Reply:
We cannot query them until the parquet file is created and added to bucket index. There is some tradeoff here.
- You can configure convert only 12h+ blocks. This has a longer delay in parquet file creation but less resource required for convertion. Users need to expect fallback to some other storage to query the data
- You can configure converting 2h block. Maybe we can configure it to only convert blocks after deduplication so parquet files are available earlier but more compactors are required to do convert

Reply:
Cool, makes more sense now. I think then parquet would need to do some kind of merge between both index for the 12h scenario.